l3kernel/l3quark.dtx

```diff
@@ -574,6 +574,291 @@
 % \end{macro}
 % \end{macro}
 %
+% \begin{macro}{\__kernel_quark_test_generate:NNNn}
+%   The function \cs{__kernel_quark_test_generate:NNNn} defines |#1|
+%   in a similar way as \cs[no-index]{quark_if_recursion_tail_stop(_do):(N|n)(n)}.
+%   using |#2| (as \cs{q_recursion_tail}) as the test quark and |#3|
+%   (as \cs{q_recursion_stop}) as the delimiter quark. |#4| is the name-space
+%   to be used (for example |__quark|).
+%
+%   There are six possible function types which this funciton can define,
+%   and which is defined depends on the signature of the function being
+%   defined:
+%   \begin{description}
+%     \item[\texttt{:n}:~]  defines \cs{quark_if_recursion_tail_stop:n}
+%     \item[\texttt{:nn}:~] defines \cs{quark_if_recursion_tail_stop_do:nn}
+%     \item[\texttt{:nN}:~] defines \cs{quark_if_recursion_tail_break:nN}
+%     \item[\texttt{:N}:~]  defines \cs{quark_if_recursion_tail_stop:N}
+%     \item[\texttt{:Nn}:~] defines \cs{quark_if_recursion_tail_stop_do:Nn}
+%     \item[\texttt{:NN}:~] defines \cs{quark_if_recursion_tail_break:NN}
+%   \end{description}
+%   Any other signature causes an error, as does a function without signature.
+%
+% \begin{macro}{\__kernel_quark_conditional_generate:NNnn}
+%   Similar to \cs{__kernel_quark_test_generate:NNNn}, but defines
+%   quark branching conditionals like \cs{quark_if_nil:nTF}
+%   using |#2| as the test quark, |#3| as the name-space to define
+%   auxiliaries, and |#4| is the list of conditional variants to define.
+%
+%   There are only two cases here:
+%   \begin{description}
+%     \item[\texttt{:n}:~] defines \cs[no-index]{quark_if_nil:n(TF)}
+%     \item[\texttt{:N}:~] defines \cs[no-index]{quark_if_nil:N(TF)}
+%   \end{description}
+%   Any other signature causes an error, as does a function without signature.
+%
+% \begin{macro}{\@@_test_generate_aux:nnNN, \@@_test_generate_choose:nN }
+%    \begin{macrocode}
+\cs_new_protected:Npn \__kernel_quark_test_generate:NNNn #1
+  {
+    \exp_last_unbraced:Nf \@@_test_generate_aux:nnNN
+      \cs_split_function:N #1
+      #1 { test }
+  }
+\cs_new_protected:Npn \__kernel_quark_conditional_generate:NNnn #1
+  {
+    \exp_last_unbraced:Nf \@@_test_generate_aux:nnNN
+      \cs_split_function:N #1
+      #1 { conditional }
+  }
+\cs_new_protected:Npn \@@_test_generate_aux:nnNN #1 #2 #3 #4
+  {
+    \reverse_if:N \tex_ifodd:D #3
+      \__kernel_msg_error:nnx { kernel } { missing-colon }
+        { \token_to_str:N #4 }
+      \exp_after:wN \use_none:nnnnnn
+    \fi:
+    \use:n { \@@_test_generate_choose:nN {#2} #4 }
+  }
+\cs_new_protected:Npn \@@_test_generate_choose:nN #1 #2 #3
+  {
+    \cs_if_exist_use:cTF { @@_#3_generate_#1:NNNn } {#2}
+      {
+        \__kernel_msg_error:nnxx { kernel } { invalid-quark-function }
+          { \token_to_str:N #2 } {#1}
+        \use_none:nnn
+      }
+  }
+%    \end{macrocode}
+% \end{macro}
+% \end{macro}
+% \end{macro}
```

Comment:
I'm not completely fixed on the “one function to define them all” idea,
but meanwhile I prefer my approach because:
 1. I think it's a good thing to have control over the name of the quarks
   used, like in `\__tl_act:NNNnn`, if I wanted to define a pair of
   quarks for a specific purpose in a package.
 2. I think it's also a good thing to have control over the name of the
   function being defined. That's why I opted to explicitly give the
   name of the function being defined, rather than generate it
   internally.

Of course, your idea can be easily extended to cover these two cases,
but then the syntax would end up being the same as mine, except that
nothing would be implied from the signature. And finally:

 3. How would your version distinguish between (say) `\quark_if_nil:n(TF)`
   and `\quark_if_nil:N(TF)`? This was the main reason I went for the
   signature-based approach: it felt really odd to have
   `\__kernel_quark_generate_conditional_n:n` and
   `\__kernel_quark_generate_conditional_N:n` as top-level functions.